lib/processor/geoip.go

```diff
@@ -0,0 +1,171 @@
+package processor
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Jeffail/benthos/v3/lib/log"
+	"github.com/Jeffail/benthos/v3/lib/metrics"
+	"github.com/Jeffail/benthos/v3/lib/types"
+	"github.com/opentracing/opentracing-go"
+	"github.com/oschwald/geoip2-golang"
+)
+
+//------------------------------------------------------------------------------
+
+func init() {
+	Constructors[TypeGeoIP] = TypeSpec{
+		constructor: NewGeoIP,
+		Description: `
+		// todo some docs
+		`,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// GeoIPConfig contains configuration fields for the GeoIP processor.
+type GeoIPConfig struct {
+	Parts        []int  `json:"parts" yaml:"parts"`
+	DatabaseType string `json:"database_type" yaml:"database_type"`
+	Database     string `json:"database" yaml:"database"`
+}
+
+// NewGeoIPConfig returns a GeoIPConfig with default values.
+func NewGeoIPConfig() GeoIPConfig {
+	return GeoIPConfig{
+		Parts:        []int{},
+		DatabaseType: "",
+		Database:     "",
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type geoipReader func(db *geoip2.Reader, ip net.IP) (interface{}, error)
+
+func cityReader() geoipReader {
+	return func(db *geoip2.Reader, ip net.IP) (interface{}, error) {
+		res, err := db.City(ip)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+}
+
+func asnReader() geoipReader {
+	return func(db *geoip2.Reader, ip net.IP) (interface{}, error) {
+		res, err := db.ASN(ip)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+}
+
+func getReader(dbtype string) (geoipReader, error) {
+	switch dbtype {
+	case "asn":
+		return asnReader(), nil
+	case "city":
+		return cityReader(), nil
+	}
+	return nil, fmt.Errorf("unknown database type: %s", dbtype)
+
+}
+
+//------------------------------------------------------------------------------
+
+// GeoIP is a processor that looks up IP addresses.
+type GeoIP struct {
+	parts   []int
+	handler *geoip2.Reader
```

Comment:
If you change this for a interface with the methods `City` and `ASN` then we can use a mock type for unit tests.

Reply:
I don't exactly understand, could you clarify a couple of things for me, please?
I could use an interface for replacing current `GeoIP.reader` an mock it but in order for that test/code to work I would be forced to explicitly declare `processor.Geoip`, not by calling `NewGeoIP`, because this would fail https://github.com/Jeffail/benthos/pull/365/files#diff-a4df8555a7687abaed0682a67d09ea3dR114
Is it ok to omit test for `NewGeoIP` or there is another way to refactor that code?

Reply:
Yeah that's fine, I'm mostly interested in testing `ProcessMessage` so I'm fine with declaring the struct instead of using `NewGeoIP`.